keps/sig-api-machinery/3716-webhook-predicates/README.md

```diff
@@ -0,0 +1,865 @@
+	// - '/' escapes to '__slash__'
+	// - Property names that exactly match a CEL RESERVED keyword escape to '__{keyword}__'. The keywords are:
+	//	  "true", "false", "null", "in", "as", "break", "const", "continue", "else", "for", "function", "if",
+	//	  "import", "let", "loop", "package", "namespace", "return".
+	// Examples:
+	//   - Expression accessing a property named "namespace": {"Expression": "object.__namespace__ > 0"}
+	//   - Expression accessing a property named "x-prop": {"Expression": "object.x__dash__prop > 0"}
+	//   - Expression accessing a property named "redact__d": {"Expression": "object.redact__underscores__d > 0"}
+	//
+	// Equality on arrays with list type of 'set' or 'map' ignores element order, i.e. [1, 2] == [2, 1].
+	// Concatenation on arrays with x-kubernetes-list-type use the semantics of the list type:
+	//   - 'set': `X + Y` performs a union where the array positions of all elements in `X` are preserved and
+	//     non-intersecting elements in `Y` are appended, retaining their partial order.
+	//   - 'map': `X + Y` performs a merge where the array positions of all keys in `X` are preserved but the values
+	//     are overwritten by values in `Y` when the key sets of `X` and `Y` intersect. Elements in `Y` with
+	//     non-intersecting keys are appended, retaining their partial order.
+	// Required.
+	Expression string `json:"expression"`
+}
+```
+
+The match condition expression is evaluated by the same libraries as those used for CEL
+ValidatingAdmissionPolicy. The only difference in expressions is the availability of the `params`
+variable. Expressions requiring access to additional information outside the AdmissionRequest must
+be performed in the webhook, and are out of scope for this proposal.
+
+### Risks and Mitigations
+
+#### Security
+
+**Risk: Attacker adds or changes a match condition to weaken an admission policy.**
+
+This is does not represent a new threat, as doing so would require update access to the admission
+registration object, and with that permission an attacker could already disable the policy through
+manipulating match rules, namespace selector, or object selector (or reroute the webhook entirely).
+
+**Risk: Logic error in match condition expression.**
+
+Currently the match conditions must be encoded in the webhook backend itself. Moving the logic
+into a CEL expression does not materially increase the risk of a logic bug.
+
+#### Debugability
+
+We do not normally log, audit, or emit an event when a webhook is out-of-scope for a request, and
+the same will _mostly_ be true for match conditions.
+
+At [log level V(5)](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-instrumentation/logging.md#what-method-to-use),
+we will emit a log when a request that would otherwise be in-scope for a webhook is excluded for a
+non-matching match condition.
+
+Short of increasing log verbosity, the recommended debug strategy is to capture or reproduce a
```

Comment:
completely outside the scope of this KEP, but a thought for future-us.  Perhaps we could create a dry-run variant that includes information about the processing chain a particular request takes for each filter and admission plugin.  That could be neat.

Reply:
Isn't this what tracing is for?

Reply:
You could take the APF strategy and just put UIDs in the response header.